flynn-host fails to fully restart after node reboot

To test automatic healing, I rebooted one node of a three-node cluster. When the node came back online, flynn-host started as expected: `sudo status flynn-host` shows 'running'. However, the local webserver on :1111 is only returning empty responses. This breaks most of the flynn-host commands available on that node and causes the cluster's discoverd process to log a continuous stream of connection failures.

On the restarted node:
```
ubuntu@ip-10-210-18-37:~$ flynn-host list
19:09:06.171373 host.go:154: Error sending HTTP request, errors: Get http://127.0.0.1:1111/services/flynn-host/instances: EOF
```

In the discoverd process logs, this message repeats several times per second:
```
[2016-09-30T19:12:30.114734548Z] 2016/09/30 19:12:30 [DEBUG] raft: Failed to contact 10.210.18.37:1111 in 2h45m59.891473048s
```

Just to be certain, I have confirmed that the restarted node is reachable by both other nodes, but attempts to `curl` the restarted node on port 1111 result in the same empty response.
```
ubuntu@ip-10-210-18-54:~$ curl http://10.210.18.37:1111/services/flynn-host/instances
curl: (52) Empty reply from server
```

--- a/discoverd/server/store.go
+++ b/discoverd/server/store.go
@@ -197,10 +197,23 @@ func (s *Store) Open() error {
 	}
 
 	// Create raft log.
+	//
+	// The mutex must be unlocked as initializing the raft store may
+	// call back into methods which acquire the lock (e.g. Restore)
+	s.mu.Unlock()
 	r, err := raft.NewRaft(config, s, cacheStore, stableStore, ss, s.peerStore, s.transport)
+	s.mu.Lock()
 	if err != nil {
 		return fmt.Errorf("raft: %s", err)
 	}
+
+	// make sure the store was not closed whilst the mutex was unlocked
+	select {
+	case <-s.closing:
+		return ErrShutdown
+	default:
+	}
+
 	s.raft = r
 
 	// Start goroutine to monitor leadership changes.
@@ -212,6 +225,10 @@ func (s *Store) Open() error {
 	go s.expirer()
 
 	return nil
+}
+
+func (s *Store) TriggerSnapshot() error {
+	return s.raft.Snapshot().Error()
 }
 
 func (s *Store) LastIndex() uint64 {

--- a/discoverd/server/store_test.go
+++ b/discoverd/server/store_test.go
@@ -11,6 +11,7 @@ import (
 
 	"github.com/flynn/flynn/discoverd/client"
 	"github.com/flynn/flynn/discoverd/server"
+	"github.com/flynn/flynn/pkg/keepalive"
 	"github.com/flynn/flynn/pkg/stream"
 )
 
@@ -682,6 +683,42 @@ func TestStore_Subscribe_NoBlock(t *testing.T) {
 	// Ensure that program does not hang.
 }
 
+// Ensure the store can be restored from a snapshot
+func TestStore_RestoreSnapshot(t *testing.T) {
+	// open a store, add some services and trigger a snapshot
+	s := MustOpenStore()
+	serviceNames := []string{"service0", "service1"}
+	for _, name := range serviceNames {
+		if err := s.AddService(name, nil); err != nil {
+			s.Close()
+			t.Fatal(err)
+		}
+	}
+	if err := s.TriggerSnapshot(); err != nil {
+		s.Close()
+		t.Fatal(err)
+	}
+	s.Store.Close()
+
+	// open another store with same path and port which will attempt
+	// to restore the snapshot
+	_, port, _ := net.SplitHostPort(s.Listener.Addr().String())
+	ln, err := keepalive.ReusableListen("tcp4", "127.0.0.1:"+port)
+	if err != nil {
+		t.Fatal(err)
+	}
+	s = NewStoreWithConfig(StoreConfig{Path: s.Path(), Listener: ln})
+	defer s.Close()
+	if err := s.Open(); err != nil {
+		t.Fatal(err)
+	}
+
+	// check the data was restored
+	if !reflect.DeepEqual(s.ServiceNames(), serviceNames) {
+		t.Fatalf("expected service names %v, got %v", serviceNames, s.ServiceNames())
+	}
+}
+
 func BenchmarkStore_AddInstance(b *testing.B) {
 	s := MustOpenStore()
 	defer s.Close()
@@ -704,25 +741,40 @@ type Store struct {
 	*server.Store
 }
 
+type StoreConfig struct {
+	Path     string
+	Listener net.Listener
+}
+
 // NewStore returns a new instance of Store.
 func NewStore() *Store {
-	// Generate a temporary path.
-	f, _ := ioutil.TempFile("", "discoverd-store-")
-	f.Close()
-	os.Remove(f.Name())
+	return NewStoreWithConfig(StoreConfig{})
+}
+
+func NewStoreWithConfig(config StoreConfig) *Store {
+	if config.Path == "" {
+		// Generate a temporary path.
+		f, _ := ioutil.TempFile("", "discoverd-store-")
+		f.Close()
+		os.Remove(f.Name())
+		config.Path = f.Name()
+	}
 
 	// Initialize store.
-	s := &Store{Store: server.NewStore(f.Name())}
+	s := &Store{Store: server.NewStore(config.Path)}
 
-	// Open listener on random port.
-	ln, err := net.Listen("tcp", "127.0.0.1:0")
-	if err != nil {
-		panic(err)
+	if config.Listener == nil {
+		// Open listener on random port.
+		ln, err := keepalive.ReusableListen("tcp4", "127.0.0.1:0")
+		if err != nil {
+			panic(err)
+		}
+		config.Listener = ln
 	}
-	_, port, _ := net.SplitHostPort(ln.Addr().String())
+	_, port, _ := net.SplitHostPort(config.Listener.Addr().String())
 
 	// Set default test settings.
-	s.Listener = ln
+	s.Listener = config.Listener
 	s.Advertise, _ = net.ResolveTCPAddr("tcp", net.JoinHostPort("localhost", port))
 	s.HeartbeatTimeout = 50 * time.Millisecond
 	s.ElectionTimeout = 50 * time.Millisecond

